### DISCUSSION TITLE
FastAPI local server: every request uses a different port -> port in use error

### DISCUSSION THREAD

**Discussion by BramVanroy** [fastapi/fastapi#9406] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
response = requests.post(rf"http://127.0.0.1:5000/endpoint1/", json={"arg1": "somevalue})
```

### Description

I have a FastAPI server running locally. I launched it with uvicorn like so:

```shell
uvicorn app:app --port 5000
```

I am querying it frpm Python, like so:

```python
response = requests.post(rf"http://127.0.0.1:5000/endpoint1/", json={"arg1": "somevalue})
```

When I am looking at my FastAPI logs, I see that every request comes from a different port and increments at every new one:

```
...
INFO:     127.0.0.1:59605 - "POST /endpoint1/ HTTP/1.1" 200 OK
INFO:     127.0.0.1:59606 - "POST /endpoint2/ HTTP/1.1" 200 OK
INFO:     127.0.0.1:59607 - "POST /endpoint1/ HTTP/1.1" 200 OK
INFO:     127.0.0.1:59608 - "POST /endpoint2/ HTTP/1.1" 200 OK
INFO:     127.0.0.1:59609 - "POST /endpoint1/ HTTP/1.1" 200 OK
INFO:     127.0.0.1:59610 - "POST /endpoint2/ HTTP/1.1" 200 OK
INFO:     127.0.0.1:59611 - "POST /endpoint1/ HTTP/1.1" 200 OK
INFO:     127.0.0.1:59612 - "POST /endpoint2/ HTTP/1.1" 200 OK
INFO:     127.0.0.1:59613 - "POST /endpoint1/ HTTP/1.1" 200 OK
INFO:     127.0.0.1:59614 - "POST /endpoint2/ HTTP/1.1" 200 OK
```

After a while this leads to issues because all ports are exhausted (I think; at port 65535). Specifically, this error:

```shell
requests.exceptions.ConnectionError: HTTPConnectionPool(host='127.0.0.1', port=5000): Max retries exceeded with url: /endpoint1/ (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x000001BC822DDC40>: Failed to establish a new connection: [WinError 10048] Only one usage of each socket address (protocol/network address/port) is normally permitted'))
```

I am doing the requests sequentially (*not* in parallel!), so is there a way to ensure that the request is always send through the same port? Or another solution to avoid this issue?

I am not sure whether this is something that relates to uvicorn and can be fixed when launching the server, or whether I need to change something in the FastAPI initialization.

### Operating System

Windows

### Operating System Details

Windows 10

### FastAPI Version

0.95.1

### Python Version

Python 3.8.10

**Comment by BramVanroy** [comment, marked as answer]
I found the issue. Python requests opens a new session on each request by default, which in turn jumps to new ports. The way around this is defining your own session

```python
session = requests.Session()
response = session.post(rf"http://127.0.0.1:5000/endpoint1/", json={"arg1": "sometext"})
# ... other requests
session.close()
```

https://stackoverflow.com/a/76046421/1150683

**Comment by Kludex** [comment]
It's not related to uvicorn neither FastAPI. It's about the requests package. Check for Session on their docs.

I can't provide links right now.
